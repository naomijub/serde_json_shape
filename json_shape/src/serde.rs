#![allow(clippy::fallible_impl_from)]
use crate::Value as JsonShape;

pub(crate) mod impls;

/// Visitor navigating a [`serde_json::Value`] based on the respective [`JsonShape`]
pub struct JsonVisitor<'json> {
    value: &'json serde_json::Value,
    shape: JsonShape,
}

impl From<serde_json::Value> for JsonShape {
    fn from(value: serde_json::Value) -> Self {
        match value {
            serde_json::Value::Null => JsonShape::Null,
            serde_json::Value::Bool(_) => JsonShape::Bool { optional: false },
            serde_json::Value::Number(_) => JsonShape::Number { optional: false },
            serde_json::Value::String(_) => JsonShape::String { optional: false },
            serde_json::Value::Array(values) => {
                if values
                    .windows(2)
                    .map(|val| {
                        (
                            JsonShape::from(val.first().unwrap()),
                            JsonShape::from(val.get(1).unwrap()),
                        )
                    })
                    .all(|val| val.0 == val.1)
                {
                    JsonShape::Array {
                        r#type: Box::new(JsonShape::from(values.first().unwrap())),
                        optional: false,
                    }
                } else {
                    JsonShape::Tuple {
                        elements: values.into_iter().map(JsonShape::from).collect(),
                        optional: false,
                    }
                }
            }
            serde_json::Value::Object(map) => JsonShape::Object {
                content: map
                    .into_iter()
                    .map(|(k, v)| (k, JsonShape::from(v)))
                    .collect(),
                optional: false,
            },
        }
    }
}

impl From<&serde_json::Value> for JsonShape {
    fn from(value: &serde_json::Value) -> Self {
        match value {
            serde_json::Value::Null => JsonShape::Null,
            serde_json::Value::Bool(_) => JsonShape::Bool { optional: false },
            serde_json::Value::Number(_) => JsonShape::Number { optional: false },
            serde_json::Value::String(_) => JsonShape::String { optional: false },
            serde_json::Value::Array(values) => {
                if values
                    .windows(2)
                    .map(|val| {
                        (
                            JsonShape::from(val.first().unwrap()),
                            JsonShape::from(val.get(1).unwrap()),
                        )
                    })
                    .all(|val| val.0 == val.1)
                {
                    JsonShape::Array {
                        r#type: Box::new(JsonShape::from(values[0].clone())),
                        optional: false,
                    }
                } else {
                    JsonShape::Tuple {
                        elements: values.iter().map(JsonShape::from).collect(),
                        optional: false,
                    }
                }
            }
            serde_json::Value::Object(map) => JsonShape::Object {
                content: map
                    .into_iter()
                    .map(|(k, v)| (k.to_string(), JsonShape::from(v)))
                    .collect(),
                optional: false,
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use serde_json::json;

    #[test]
    fn test_from_json_null() {
        let json = json!(null);
        let ref_shape = JsonShape::from(&json);
        let shape = JsonShape::from(json);
        assert_eq!(shape, JsonShape::Null);
        assert_eq!(ref_shape, JsonShape::Null);
    }

    #[test]
    fn test_from_json_number() {
        let json = json!(123.456);
        let ref_shape = JsonShape::from(&json);
        let shape = JsonShape::from(json);
        assert_eq!(shape, JsonShape::Number { optional: false });
        assert_eq!(ref_shape, JsonShape::Number { optional: false });
    }

    #[test]
    fn test_from_json_string() {
        let json = json!("string");
        let ref_shape = JsonShape::from(&json);
        let shape = JsonShape::from(json);
        assert_eq!(shape, JsonShape::String { optional: false });
        assert_eq!(ref_shape, JsonShape::String { optional: false });
    }

    #[test]
    fn test_from_json_bool() {
        let json = json!(true);
        let ref_shape = JsonShape::from(&json);
        let shape = JsonShape::from(json);
        assert_eq!(shape, JsonShape::Bool { optional: false });
        assert_eq!(ref_shape, JsonShape::Bool { optional: false });
    }

    #[test]
    fn test_from_json_array() {
        let json = json!([1, 2, 3]);
        let ref_shape = JsonShape::from(&json);
        let shape = JsonShape::from(json);
        assert_eq!(
            shape,
            JsonShape::Array {
                r#type: Box::new(JsonShape::Number { optional: false }),
                optional: false
            }
        );
        assert_eq!(
            ref_shape,
            JsonShape::Array {
                r#type: Box::new(JsonShape::Number { optional: false }),
                optional: false
            }
        );
    }

    #[test]
    fn test_from_json_tuple() {
        let json = json!([1, "string", true, null]);
        let ref_shape = JsonShape::from(&json);
        let shape = JsonShape::from(json);
        assert_eq!(
            shape,
            JsonShape::Tuple {
                elements: vec![
                    JsonShape::Number { optional: false },
                    JsonShape::String { optional: false },
                    JsonShape::Bool { optional: false },
                    JsonShape::Null
                ],
                optional: false
            }
        );
        assert_eq!(
            ref_shape,
            JsonShape::Tuple {
                elements: vec![
                    JsonShape::Number { optional: false },
                    JsonShape::String { optional: false },
                    JsonShape::Bool { optional: false },
                    JsonShape::Null
                ],
                optional: false
            }
        );
    }

    #[test]
    fn test_from_json_object() {
        let json = json!({
            "a": 1,
            "b": "string",
            "c": [12, 34, 56],
            "d": null,
            "e": true,
            "f": [12, true, "string"],
            "objs": [
                {
                    "a": 1,
                    "b" : "string"
                },
                {
                    "a": 2,
                    "c" : true
                }
            ],
            "obj": {
                "d": 1,
                "e" : "string"
            }
        });
        let shape = JsonShape::from(json);
        assert_snapshot!(shape.to_string());
    }
}
