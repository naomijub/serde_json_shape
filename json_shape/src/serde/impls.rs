use crate::{serde::JsonVisitor, value::Value as JsonShape};

impl<'json> From<&'json serde_json::Value> for JsonVisitor<'json> {
    fn from(value: &'json serde_json::Value) -> Self {
        let shape = JsonShape::from(value);
        Self { value, shape }
    }
}

impl JsonVisitor<'_> {
    /// Returns the original [`serde_json::Value`]
    #[must_use]
    pub const fn value(&self) -> &serde_json::Value {
        self.value
    }

    /// Returns the [`JsonShape`] of the original [`serde_json::Value`]
    #[must_use]
    pub const fn shape(&self) -> &JsonShape {
        &self.shape
    }
}
