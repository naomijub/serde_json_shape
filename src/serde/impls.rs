use crate::{serde::JsonVisitor, value::Value as JsonShape};

impl<'json> From<&'json serde_json::Value> for JsonVisitor<'json> {
    fn from(value: &'json serde_json::Value) -> Self {
        let shape = JsonShape::from(value);
        Self { value, shape }
    }
}

impl JsonVisitor<'_> {
    /// Returns the original [`serde_json::Value`]
    #[must_use]
    pub const fn value(&self) -> &serde_json::Value {
        self.value
    }

    /// Returns the [`JsonShape`] of the original [`serde_json::Value`]
    #[must_use]
    pub const fn shape(&self) -> &JsonShape {
        &self.shape
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn from_value_null() {
        let value = serde_json::Value::Null;
        let visitor = JsonVisitor::from(&value);
        assert_eq!(visitor.value(), &value);
        assert_eq!(visitor.shape(), &JsonShape::Null);
    }

    #[test]
    fn from_value_number() {
        let value = serde_json::Value::Number(3.into());
        let visitor = JsonVisitor::from(&value);
        assert_eq!(visitor.value(), &value);
        assert_eq!(visitor.shape(), &JsonShape::Number { optional: false });
    }
}
