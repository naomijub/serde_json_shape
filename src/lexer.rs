#![cfg(not(tarpaulin_include))]
#![allow(clippy::range_plus_one)]
#![allow(clippy::unnecessary_struct_initialization)]
use std::fmt;

use super::parser::{Diagnostic, Span};
use codespan_reporting::diagnostic::Label;
use logos::{Lexer, Logos};

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub enum LexerError {
    #[default]
    Invalid,
    UnterminatedString,
}

impl LexerError {
    pub fn into_diagnostic(self, span: Span) -> Diagnostic {
        match self {
            Self::Invalid => Diagnostic::error()
                .with_message("invalid token")
                .with_labels(vec![Label::primary((), span)]),
            Self::UnterminatedString => Diagnostic::error()
                .with_message("unterminated string")
                .with_labels(vec![Label::primary((), span)]),
        }
    }
}

#[cfg(not(tarpaulin_include))]
fn parse_string(lexer: &mut Lexer<'_, Token>) -> Result<(), LexerError> {
    let mut it = lexer.remainder().chars();
    while let Some(c) = it.next() {
        match c {
            '"' => {
                lexer.bump(1);
                return Ok(());
            }
            '\\' => {
                lexer.bump(1);
                if let Some(c) = it.next() {
                    lexer.bump(c.len_utf8());
                }
            }
            c => {
                lexer.bump(c.len_utf8());
            }
        }
    }
    Err(LexerError::UnterminatedString)
}

#[cfg(not(tarpaulin_include))]
fn check_string(value: &str, span: &Span, diags: &mut Vec<Diagnostic>) {
    let mut it = value.chars().enumerate();
    while let Some((i, c)) = it.next() {
        match c {
            '\\' => match it.next() {
                Some((_, '"' | '\\' | '/' | 'b' | 'f' | 'n' | 'r' | 't')) => {}
                Some((i, 'u')) => {
                    for j in 0..4 {
                        if !it.next().is_some_and(|(_, c)| c.is_ascii_hexdigit()) {
                            diags.push(
                                Diagnostic::error()
                                    .with_message("invalid unicode escape sequence")
                                    .with_labels(vec![Label::primary(
                                        (),
                                        (span.start + i - 1)..(span.start + i + j + 1),
                                    )]),
                            );
                            break;
                        }
                    }
                }
                Some((j, _)) => {
                    diags.push(
                        Diagnostic::error()
                            .with_message("invalid escape sequence")
                            .with_labels(vec![Label::primary(
                                (),
                                span.start + j - 1..span.start + j + 1,
                            )]),
                    );
                }
                _ => unreachable!(),
            },
            '\u{0020}'..='\u{10FFFF}' => {}
            c => {
                diags.push(
                    Diagnostic::error()
                        .with_message(format!("string contains invalid character {c:?}"))
                        .with_labels(vec![
                            Label::primary((), span.start + i..span.start + i + 1)
                                .with_message("after this character"),
                        ]),
                );
            }
        }
    }
}

// TODO: implement lexer
#[allow(clippy::upper_case_acronyms)]
#[derive(Logos, Debug, PartialEq, Eq, Copy, Clone)]
#[logos(error = LexerError)]
pub enum Token {
    EOF,
    #[regex("[\u{0020}\u{0009}]+")]
    Whitespace,
    #[regex("\r?\n")]
    Newline,
    #[token("true")]
    True,
    #[token("false")]
    False,
    #[token("null")]
    Null,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token("[")]
    LBrak,
    #[token("]")]
    RBrak,
    #[token(",")]
    Comma,
    #[token(":")]
    Colon,
    #[regex("\"", parse_string)]
    String,
    #[regex(r"-?(0|[1-9][0-9]*)(\.[0-9]+)?([eE][+-]?[0-9]+)?")]
    Number,
    #[regex(r"[a-zA-Z][a-zA-Z0-9]*", |_| false)]
    Error,
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Token::EOF => "SYSNULL",
                Token::Newline | Token::Whitespace => "",
                Token::False | Token::True => "Boolean",
                Token::Null => "Null",
                Token::LBrace => "{",
                Token::RBrace => "}",
                Token::LBrak => "[",
                Token::RBrak => "]",
                Token::Comma => ",",
                Token::Colon => ":",
                Token::String => "String",
                Token::Number => "Number",
                Token::Error => "Unknown Error",
            }
        )
    }
}

pub fn tokenize(source: &str, diags: &mut Vec<Diagnostic>) -> (Vec<Token>, Vec<Span>) {
    let lexer = Token::lexer(source);
    let mut tokens = vec![];
    let mut spans = vec![];
    let source = lexer.source();

    let mut count_brace = 0;
    let mut count_brak = 0;
    for (token, span) in lexer.spanned() {
        match token {
            Ok(token) => {
                match token {
                    Token::String => {
                        check_string(&source[span.start..span.end], &span, diags);
                    }
                    Token::LBrace => count_brace += 1,
                    Token::RBrace => count_brace -= 1,
                    Token::LBrak => count_brak += 1,
                    Token::RBrak => count_brak -= 1,
                    _ => {}
                }
                if count_brace + count_brak > 256 {
                    diags.push(
                        Diagnostic::error()
                            .with_message("bracket nesting level exceeded maximum of 256")
                            .with_labels(vec![Label::primary((), span)]),
                    );
                    break;
                }
                tokens.push(token);
            }
            Err(err) => {
                diags.push(err.into_diagnostic(span.clone()));
                tokens.push(Token::Error);
            }
        }
        spans.push(span);
    }
    (tokens, spans)
}
