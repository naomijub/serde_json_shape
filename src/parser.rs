// Mostly codegen file
#![cfg(not(tarpaulin_include))]
#![allow(clippy::pedantic)]
#![allow(clippy::nursery)]
use super::lexer::{Token, tokenize};
use codespan_reporting::diagnostic::Label;
use serde::Serialize;

pub type Diagnostic = codespan_reporting::diagnostic::Diagnostic<()>;

#[derive(Default)]
pub struct Context<'a> {
    marker: std::marker::PhantomData<&'a ()>,
}

include!(concat!(env!("OUT_DIR"), "/generated.rs"));

impl ParserCallbacks for Parser<'_> {
    fn create_tokens(source: &str, diags: &mut Vec<Diagnostic>) -> (Vec<Token>, Vec<Span>) {
        tokenize(source, diags)
    }
    fn create_diagnostic(&self, span: Span, message: String) -> Diagnostic {
        Diagnostic::error()
            .with_message(message)
            .with_labels(vec![Label::primary((), span)])
    }
}

impl PartialEq for CstIndex {
    fn eq(&self, other: &Self) -> bool {
        self.0 == other.0
    }
}

impl PartialEq for Node {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::Rule(l0, l1), Self::Rule(r0, r1)) => l0 == r0 && l1 == r1,
            (Self::Token(l0, l1), Self::Token(r0, r1)) => l0 == r0 && l1 == r1,
            _ => false,
        }
    }
}

impl PartialEq for Cst<'_> {
    fn eq(&self, other: &Self) -> bool {
        self.source == other.source
            && self.spans == other.spans
            && self.nodes == other.nodes
            && self.token_count == other.token_count
            && self.non_skip_len == other.non_skip_len
    }
}

impl std::fmt::Debug for Cst<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Cst")
            .field("source", &self.source)
            .field("spans", &self.spans)
            .field("nodes", &self.nodes)
            .field("token_count", &self.token_count)
            .field("non_skip_len", &self.non_skip_len)
            .finish()
    }
}

impl Serialize for Cst<'_> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(self.source)
    }
}

#[cfg(test)]
mod tests {
    use super::Parser;
    use insta::assert_snapshot;

    #[test]
    fn parse_json() {
        let json = r#"{
            "str": "this is a string",
            "number": 123.456,
            "bool_true": true,
            "bool_false": false,
            "nil": null,
            "array": [123, "string", true],
            "map": {
            "a": "b",
            "c": 123
            },
            "array of maps": [
                {
                    "a": "b",
                    "c": 123
                },
                {
                    "a": "b",
                    "b": true
                }
            ]
        }"#;

        let mut diags = Vec::new();
        let parsed = Parser::parse(json, &mut diags);

        assert_snapshot!(parsed);
    }
}
