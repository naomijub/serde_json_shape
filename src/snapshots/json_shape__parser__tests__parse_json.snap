---
source: src/parser.rs
expression: parsed
---
file [0..530]
    object [0..530]
        LBrace "{" [0..1]
        Newline "\n" [1..2]
        Whitespace "            " [2..14]
        member [14..39]
            String "\"str\"" [14..19]
            Colon ":" [19..20]
            Whitespace " " [20..21]
            literal [21..39]
                String "\"this is a string\"" [21..39]
        Comma "," [39..40]
        Newline "\n" [40..41]
        Whitespace "            " [41..53]
        member [53..70]
            String "\"number\"" [53..61]
            Colon ":" [61..62]
            Whitespace " " [62..63]
            literal [63..70]
                Number "123.456" [63..70]
        Comma "," [70..71]
        Newline "\n" [71..72]
        Whitespace "            " [72..84]
        member [84..101]
            String "\"bool_true\"" [84..95]
            Colon ":" [95..96]
            Whitespace " " [96..97]
            literal [97..101]
                boolean [97..101]
                    True "true" [97..101]
        Comma "," [101..102]
        Newline "\n" [102..103]
        Whitespace "            " [103..115]
        member [115..134]
            String "\"bool_false\"" [115..127]
            Colon ":" [127..128]
            Whitespace " " [128..129]
            literal [129..134]
                boolean [129..134]
                    False "false" [129..134]
        Comma "," [134..135]
        Newline "\n" [135..136]
        Whitespace "            " [136..148]
        member [148..159]
            String "\"nil\"" [148..153]
            Colon ":" [153..154]
            Whitespace " " [154..155]
            literal [155..159]
                Null "null" [155..159]
        Comma "," [159..160]
        Newline "\n" [160..161]
        Whitespace "            " [161..173]
        member [173..203]
            String "\"array\"" [173..180]
            Colon ":" [180..181]
            Whitespace " " [181..182]
            array [182..203]
                LBrak "[" [182..183]
                literal [183..186]
                    Number "123" [183..186]
                Comma "," [186..187]
                Whitespace " " [187..188]
                literal [188..196]
                    String "\"string\"" [188..196]
                Comma "," [196..197]
                Whitespace " " [197..198]
                literal [198..202]
                    boolean [198..202]
                        True "true" [198..202]
                RBrak "]" [202..203]
        Comma "," [203..204]
        Newline "\n" [204..205]
        Whitespace "            " [205..217]
        member [217..282]
            String "\"map\"" [217..222]
            Colon ":" [222..223]
            Whitespace " " [223..224]
            object [224..282]
                LBrace "{" [224..225]
                Newline "\n" [225..226]
                Whitespace "            " [226..238]
                member [238..246]
                    String "\"a\"" [238..241]
                    Colon ":" [241..242]
                    Whitespace " " [242..243]
                    literal [243..246]
                        String "\"b\"" [243..246]
                Comma "," [246..247]
                Newline "\n" [247..248]
                Whitespace "            " [248..260]
                member [260..268]
                    String "\"c\"" [260..263]
                    Colon ":" [263..264]
                    Whitespace " " [264..265]
                    literal [265..268]
                        Number "123" [265..268]
                Newline "\n" [268..269]
                Whitespace "            " [269..281]
                RBrace "}" [281..282]
        Comma "," [282..283]
        Newline "\n" [283..284]
        Whitespace "            " [284..296]
        member [296..520]
            String "\"array of maps\"" [296..311]
            Colon ":" [311..312]
            Whitespace " " [312..313]
            array [313..520]
                LBrak "[" [313..314]
                Newline "\n" [314..315]
                Whitespace "                " [315..331]
                object [331..409]
                    LBrace "{" [331..332]
                    Newline "\n" [332..333]
                    Whitespace "                    " [333..353]
                    member [353..361]
                        String "\"a\"" [353..356]
                        Colon ":" [356..357]
                        Whitespace " " [357..358]
                        literal [358..361]
                            String "\"b\"" [358..361]
                    Comma "," [361..362]
                    Newline "\n" [362..363]
                    Whitespace "                    " [363..383]
                    member [383..391]
                        String "\"c\"" [383..386]
                        Colon ":" [386..387]
                        Whitespace " " [387..388]
                        literal [388..391]
                            Number "123" [388..391]
                    Newline "\n" [391..392]
                    Whitespace "                " [392..408]
                    RBrace "}" [408..409]
                Comma "," [409..410]
                Newline "\n" [410..411]
                Whitespace "                " [411..427]
                object [427..506]
                    LBrace "{" [427..428]
                    Newline "\n" [428..429]
                    Whitespace "                    " [429..449]
                    member [449..457]
                        String "\"a\"" [449..452]
                        Colon ":" [452..453]
                        Whitespace " " [453..454]
                        literal [454..457]
                            String "\"b\"" [454..457]
                    Comma "," [457..458]
                    Newline "\n" [458..459]
                    Whitespace "                    " [459..479]
                    member [479..488]
                        String "\"b\"" [479..482]
                        Colon ":" [482..483]
                        Whitespace " " [483..484]
                        literal [484..488]
                            boolean [484..488]
                                True "true" [484..488]
                    Newline "\n" [488..489]
                    Whitespace "                " [489..505]
                    RBrace "}" [505..506]
                Newline "\n" [506..507]
                Whitespace "            " [507..519]
                RBrak "]" [519..520]
        Newline "\n" [520..521]
        Whitespace "        " [521..529]
        RBrace "}" [529..530]
